### ISSUE TITLE
Git seems to be broken for Fish shell on MacOS

### ISSUE THREAD

**Issue by joedborg** [justjanne/powerline-go#46]
I use fish as my main shell.  After doing an update (with `go get -u`), the git integration seems to have stopped working.  By that I mean it doesn't show up with the branch, changes etc., but powerline-go is running.

If I fire bash up instead, it is working. 

I'm running fish 2.6.0, git 2.14.3 and latest pull of powerline-go.

Please let me know if I can give any more information.

Thanks

**Comment by joedborg** [comment]
Something I've noticed:

```
 joe  wallace  project  $  
 joe  wallace  project  $  /Users/joe/Library/Go/bin/powerline-go -shell bare
 joe  wallace  Projects  project  1.0.0-devel…  $  ⏎                                                                            
 joe  wallace  project  $  
```

Fish config is:
```
function fish_prompt
    /Users/joe/Library/Go/bin/powerline-go -error $status -shell bare
end
```

**Comment by justjanne** [comment]
Can you try running

    /Users/joe/Library/Go/bin/powerline-go -shell bare -modules git

And tell me if that correctly shows the git module?

**Comment by joedborg** [comment]
Hey @justjanne, thanks for the quick reply.  When I do that, I get no PS1 at all (but can still issue commands).

**Comment by justjanne** [comment]
No, I meant, running that in a folder that has a git repo as a normal command — not setting it as PS1

**Comment by joedborg** [comment]
Ah, apologies.  If I do that, I get the branch and changes.

**Comment by justjanne** [comment]
Good, so the module works. Now we just need to find out why it doesn't work in your shell.

**Comment by joedborg** [comment]
Is there any way to get some debug output?

**Comment by justjanne** [comment]
Not really, but I'll investigate this further today.

**Comment by joedborg** [comment]
Thanks!  Please let me know if I can help in any way.

**Comment by justjanne** [comment]
So, it seems to work for me with this workaround:

```
function fish_prompt
    export real_status=$status
    export COLUMNS=$COLUMNS
    ~/go/bin/powerline-go -error $real_status -shell bare
end
```

The issue seems to be that fish on macOS for some reason doesn’t export the COLUMNS variable, and so the shell thinks it’s being run in a tiny terminal

**Comment by joedborg** [comment]
Awesome, that works, thanks!

Not sure what's changed in terms of fish or macOS or why updating powerline triggered that.

Thanks again for finding a fix :)

**Comment by rocketraman** [comment]
NOTE, I have the same problem here with fish 2.3.1 on Linux (Fedora 26). In my case also, running `powerline-go -shell bare -modules git` in a git directory works correctly. However, when run "normally" I don't always get branch information. Though some branches like `master` *do* work, so it seems to be related to length somehow.

I tried the workaround with COLUMNS above, but my version of `fish` does not allow it to be modified:

`set: Tried to change the read-only variable “COLUMNS”`

**Comment by justjanne** [comment]
@rocketraman In that case, can you try adding -max-width 0? That'll disable the scaling.

**Comment by rocketraman** [comment]
> @rocketraman In that case, can you try adding -max-width 0? That'll disable the scaling.

@justjanne That fixed it! Is this an issue with how `fish` reports terminal size to `powerline-go`? Or with how `powerline-go` is interpreting it?

**Comment by justjanne** [comment]
It seems to be an issue with how each shell is reporting terminal size.

Sadly, if the prompt gets too long, fish just removes it entirely.

**Comment by rocketraman** [comment]
> Sadly, if the prompt gets too long, fish just removes it entirely.

Just to clarify, fish was not removing the prompt entirely -- only parts of it (the full path and branch info) were being truncated i.e. before:

```
 raman  edison  some-projec…  $ 
```
and after `-max-width 0`

```
 raman  edison  ~  source  some-project-name  proposed  $  
```

**Comment by rocketraman** [comment]
> Sadly, the columns variable is often exported wrongly, so it's mostly broken.

To test this, I set my prompt to:

```
function fish_prompt
    echo "$COLUMNS > "
end
```

I got:

```
227 > 
```

which seems to be correct.

Are you sure powerline-go is doing the right thing here?

**Comment by justjanne** [comment]
Considering it works fine in zsh, where I use it every day, I'm quite sure it's doing the right thing

**Comment by justjanne** [comment]
It also works fine in bash.

**Comment by justjanne** [comment]
@rocketraman I’ve made a workaround that (a) fixes the terminal size detection via the ioctl syscall, so we don’t have to rely on $COLUMNS anymore, and (b), if $COLUMNS isn’t found, and the syscall doesn’t work, we simply print the entire prompt now.

**Comment by rocketraman** [comment]
> @rocketraman I’ve made a workaround that (a) fixes the terminal size detection via the ioctl syscall, so we don’t have to rely on $COLUMNS anymore, and (b), if $COLUMNS isn’t found, and the syscall doesn’t work, we simply print the entire prompt now.

Ooh nice! Doesn't sound like a "work-around" : sounds like the right way to do things :-) The only down-side is we won't be able to use powerline-go as a useful test case to figure out what's going on in `fish` :-).
